primitives/npos-elections/src/lib.rs

```diff
@@ -629,32 +629,37 @@ pub(crate) fn setup_inputs<AccountId: IdentifierT>(
 		})
 		.collect::<Vec<CandidatePtr<AccountId>>>();
 
-	let voters = initial_voters.into_iter().map(|(who, voter_stake, votes)| {
+	let voters = initial_voters.into_iter().filter_map(|(who, voter_stake, votes)| {
 		let mut edges: Vec<Edge<AccountId>> = Vec::with_capacity(votes.len());
-		for v in votes {
-			if edges.iter().any(|e| e.who == v) {
-				// duplicate edge.
-				continue;
-			}
-			if let Some(idx) = c_idx_cache.get(&v) {
-				// This candidate is valid + already cached.
-				let mut candidate = candidates[*idx].borrow_mut();
-				candidate.approval_stake =
-					candidate.approval_stake.saturating_add(voter_stake.into());
-				edges.push(
-					Edge {
-						who: v.clone(),
-						candidate: Rc::clone(&candidates[*idx]),
-						..Default::default()
-					}
-				);
-			} // else {} would be wrong votes. We don't really care about it.
+		if edges.is_empty() {
```

Comment:
This is always going to be empty! you just build a new `edges` and you immediately assert that it is empty. 

You can't do this like such. You need to go over all `voters`, some might be duplicate, some might also be invalid  
(`// else {} would be wrong votes. We don't really care about it.`). Once all is done, before returning `Voter { .. } ` you check if it has any edges or not. 

Also, a test would be good.

Reply:
Okay ... got it